routes/api/users.js

```diff
@@ -1,10 +1,75 @@
 //Handle registering users, adding users, etc.
 const express = require('express');
 const router = express.Router();
+const gravatar = require('gravatar');
+const bcrypt = require('bcryptjs');
+const { check, validationResult } = require('express-validator/check');
 
-// @route   GET api/users
-// @desc    Test route
+const User = require('../../models/User');
+
+// @route   POST api/users
+// @desc    Register user
 // @access  Public
-router.get('/', (req, res) => res.send('User route'));
+router.post(
+  '/',
+  [
+    check('name', 'Name is required!')
+      .not()
+      .isEmpty(),
+    check('email', 'Please include a valid email').isEmail(),
+    check(
+      'password',
+      'Please enter a password with 8 or more characters'
+    ).isLength({ min: 6 })
+  ],
+  async (req, res) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      //If they don't include above information correctly, then this a bad request.
+      return res.status(400).json({ errors: errors.array() });
+    }
+
+    const { name, email, password } = req.body;
+
+    try {
+      //Check if the user exits
+      let user = await User.findOne({ email });
```

Comment:
We have a unique index on email field in MongoDB. This means if a duplicate email tries to be inserted, the insertion will fail.

You can instead of querying for the email to check yourself, just perform the insert and take the result from that insert. You will probably have to double check what it returns in express, but it will probably be 1 or 0 (1 for success, 0 for fail). If it fails, then you know theres an email already in use.

I think it would be better than a search query, because you would not have to retrieve the entire document from DB.

Reply:
Actually, rethinking this, it may be better to use find() because we can check deleted_flag...

But that also depends on what we want behavior to be when an account is deleted. Do we want the user to not use the same email to make a new account if they wish (and instead be offered the opportunity to "reactivate" their account?) This would probably need a little discussion with the team.

For now, though, I think we can ignore deletion functionality and take it one step at a time.